### PR TITLE
Migrate datatreee assertions/extensions/formatting

### PR DESCRIPTION
**Pull request by owenlittlejohns**
This PR continues the overall work of [migrating DataTree into xarray](https://github.com/pydata/xarray/issues/8572).

* `xarray/core/datatree_render.py` is the renamed version of `xarray/datatree_/datatree/render.py`.
* `xarray/core/extensions.py` now contains functionality from `xarray/datatree_/datatree/extensions.py`.
* `xarray/core/formatting.py` now contains functionality from `xarray/datatree_/datatree/formatting.py`.
* `xarray/tests/test_datatree.py` now contains tests from `xarray/datatree_/datatree/tests/test_dataset_api.py`.
* `xarray/testing/assertions.py` now contains functionality from `/xarray/datatree_/datatree/testing.py`.

I had also meant to get to `common.py` and what's left of `io.py`, but I've got a hefty couple of days of meetings ahead, so I wanted to get this progress into PR before that happens. @flamingbear or I can follow up with the remaining things in a separate PR. (Also this PR is already getting a little big, so maybe it's already got enough in it)

- [x] Contributes to migration step for miscellaneous modules in #8572
- [ ] ~~Tests added~~
- [ ] ~~User visible changes (including notable bug fixes) are documented in `whats-new.rst`~~
- [ ] ~~New functions/methods are listed in `api.rst`~~

### CODE REVIEW

File: xarray/core/datatree_mapping.py [hunk 1]
```diff
@@ -71,37 +71,6 @@ def check_isomorphic(
         raise TreeIsomorphismError("DataTree objects are not isomorphic:\n" + diff)
 
 
-def diff_treestructure(a: DataTree, b: DataTree, require_names_equal: bool) -> str:
```
Comment (owenlittlejohns) [inline xarray/core/datatree_mapping.py:74]: I moved this into `xarray/core/formatting.py` to avoid a circular dependency issue.

File: xarray/core/datatree_render.py [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+            maxlevel: Limit rendering to this depth.
+        :any:`RenderDataTree` is an iterator, returning a tuple with 3 items:
+        `pre`
+            tree prefix.
+        `fill`
+            filling for multiline entries.
+        `node`
+            :any:`NodeMixin` object.
+        It is up to the user to assemble these parts to a whole.
+        >>> from xarray import Dataset
```
Comment (owenlittlejohns) [inline xarray/core/datatree_render.py:94]: The examples in this documentation string are a bit shorter than the originals from `anytree`. That's because using `node.children.values()` gets a `ValuesView` which isn't compatible with iterables like `reversed` that alter the order of the items in the iterable.

File: xarray/datatree_/datatree/ops.py [hunk 3]
```diff
@@ -1,6 +1,6 @@
 import textwrap
 
-from xarray import Dataset
+from xarray.core.dataset import Dataset
```
Comment (owenlittlejohns) [inline xarray/datatree_/datatree/ops.py:3]: This was causing another circular dependency issue. @flamingbear - just FYI, for when you are tweaking `ops.py`.

File: xarray/testing/__init__.py [hunk 4]
```diff
@@ -1,3 +1,4 @@
+# TODO: Add assert_isomorphic when making DataTree API public
```
Comment (owenlittlejohns) [inline xarray/testing/__init__.py:1]: I assumed we didn't want to surface `assert_isomorphic` until everything else was public. Does that sound good to you @TomNicholas?

File: xarray/testing/assertions.py [hunk 5]
```diff
@@ -81,12 +134,19 @@ def assert_equal(a, b):
         assert a.equals(b), formatting.diff_dataset_repr(a, b, "equals")
     elif isinstance(a, Coordinates):
         assert a.equals(b), formatting.diff_coords_repr(a, b, "equals")
+    elif isinstance(a, DataTree):
+        from_root = kwargs.get("from_root", True)
```
Comment (owenlittlejohns) [inline xarray/testing/assertions.py:138]: Would people be more comfortable if I was doing some type checking here for `from_root` being a `bool`?

File: xarray/tests/test_extensions.py [hunk 6]
```diff
@@ -5,9 +5,14 @@
 import pytest
 
 import xarray as xr
+
+# TODO: Remove imports in favour of xr.DataTree etc, once part of public API
```
Comment (owenlittlejohns) [inline xarray/tests/test_extensions.py:9]: I'm also hoping that once we can for `xr.DataTree` that all the type annotations that `mypy` has insisted on in the tests might be able to be removed (e.g., `dt: DataTree = DataTree()`)

File: xarray/tests/test_datatree.py [hunk 7]
```diff
@@ -729,4 +728,99 @@ def test_filter(self):
             name="Abe",
         )
         elders = simpsons.filter(lambda node: node["age"].item() > 18)
-        dtt.assert_identical(elders, expected)
+        assert_identical(elders, expected)
+
+
+class TestDSMethodInheritance:
```
Comment (owenlittlejohns) [inline xarray/tests/test_datatree.py:734]: This felt like a good place for adding these tests. But I could move them to another module if that's a preference (just probably not under the name `test_dataset_api.py`, which might be confusing outside of a specifically `datatree` oriented context).

File: xarray/testing/assertions.py [hunk 8]
```diff
@@ -50,21 +52,72 @@ def _data_allclose_or_equiv(arr1, arr2, rtol=1e-05, atol=1e-08, decode_bytes=Tru
...
     objects.
 
     Raises an AssertionError if two objects are not equal. This will match
     data values, dimensions and coordinates, but not names or attributes
     (except for Dataset objects for which the variable names must match).
-    Arrays with NaN in the same location are considered equal.
+    Arrays with NaN in the same location are considered equal. DataTrees are
+    equal if they have isomorphic node structures, with matching node names,
+    and if they have matching variables and coordinates, all of which are
+    equal.
```
Comment (TomNicholas) [inline xarray/testing/assertions.py:109]: The definition here of `equal/allclose/identical` is "isomorphic, and also every set of corresponding nodes in the trees are equal/allclose/identical". It would be good to reword this to indicate that `assert_equal` for trees is essentially just `assert_equal` for `Dataset` but mapped over the trees.

File: xarray/testing/assertions.py [hunk 9]
```diff
@@ -50,21 +52,72 @@ def _data_allclose_or_equiv(arr1, arr2, rtol=1e-05, atol=1e-08, decode_bytes=Tru
...
-    b : xarray.Dataset, xarray.DataArray, xarray.Variable or xarray.Coordinates
-        The second object to compare.
+    a : xarray.Dataset, xarray.DataArray, xarray.Variable, xarray.Coordinates
+        or xarray.core.datatree.DataTree. The first object to compare.
+    b : xarray.Dataset, xarray.DataArray, xarray.Variable, xarray.Coordinates
+        or xarray.core.datatree.DataTree. The second object to compare.
+    **kwargs : dict
+        Additional keyword arguments passed on to DataTree.equals. Specifically,
+        from_root, which defaults True, and will check the whole tree above
+        the given node.
```
Comment (TomNicholas) [inline xarray/testing/assertions.py:120]: I don't understand why `assert_equal` has `**kwargs` at all - they don't seem to be used?

`assert_allclose` accepts specific named tolerance kwargs, shouldn't we just add `from_root` explicitly to all of these functions?

Comment (owenlittlejohns) [reply xarray/testing/assertions.py:120]: I may have misunderstood the conversation in our tag-up today, but it sounded like `**kwargs` was a potentially preferred option to explicitly stating `from_root`, where `from_root` is only used by `assert_equal` or `assert_identical` when dealing with a `DataTree`. But happy to explicitly name `from_root` as an optional argument if that seems better.

Comment (TomNicholas) [reply xarray/testing/assertions.py:120]: Forgetting about `from_root` for a moment, I just don't understand why xarray's existing `assert_equal` method already accepts `**kwargs` then does nothing with them.

Comment (owenlittlejohns) [reply xarray/testing/assertions.py:120]: I'm not sure I follow here - I added `**kwargs` to both `assert_equal` and `assert_identical` in this PR. Then inside the condition for `isinstance(a, DataTree)` I try to grab `from_root` from those `kwargs`.

Comment (TomNicholas) [reply xarray/testing/assertions.py:120]: Oops, I'm being stupid, ignore the comment about it already being there (it wasn't).

Why are you adding `**kwargs` though? Did we decide to do that in the meeting? I thought all we decided was that `assert_*` would need a `from_root` arg, even if that didn't do anything for non-`DataTree` inputs.

Comment (owenlittlejohns) [reply xarray/testing/assertions.py:120]: Okay. I can make that change, sounds good.

Comment (owenlittlejohns) [reply xarray/testing/assertions.py]: [63e2836](https://github.com/pydata/xarray/pull/8967/commits/63e28362f22881865d2f0c077d02506675d3630a)

Comment (keewis) [reply xarray/testing/assertions.py]: I've been suggesting that... the idea was that having it for non-`DataTree` objects might be confusing, but I guess `**kwargs` makes tab-completion less useful (but other than that there should not be any difference).

Comment (TomNicholas) [reply xarray/testing/assertions.py]: Ah my bad for somehow missing that suggestion initially.

Still, I feel like hiding a kwarg away is more confusing than just showing it and saying it is only used if two `DataTree` objects are passed... For example [`apply_ufunc`](https://docs.xarray.dev/en/stable/generated/xarray.apply_ufunc.html#xarray.apply_ufunc) has arguments that are only used if the input is of a certain type.

Another thing we could do here is use `typing.overloads` to ensure that passing `from_root` when `a` and `b` are not both `DataTree`s would fail static type checking.

Comment (owenlittlejohns) [reply xarray/testing/assertions.py]: Yeah - I was just thinking about using `typing.overloads` - I'll have a go at that.

File: xarray/testing/assertions.py [hunk 10]
```diff
@@ -50,7 +52,59 @@ def _data_allclose_or_equiv(arr1, arr2, rtol=1e-05, atol=1e-08, decode_bytes=Tru
...
+
+        assert a.isomorphic(b, from_root=from_root), diff_datatree_repr(
+            a, b, "isomorphic"
+        )
+    else:
+        raise TypeError(f"{type(a)} not of type DataTree")
+
+
+@overload
+def assert_equal(a, b): ...
```
Comment (owenlittlejohns) [inline xarray/testing/assertions.py:99]: I initially tried to specify all the individual overloads (e.g., `assert_equal(a: Dataset, b: Dataset)`, etc). That led to issues with the return values from `DataTree.__getitem__`, which are: [DataTree | DataArray](https://github.com/pydata/xarray/blob/main/xarray/core/datatree.py#L848).

Hopefully this hits enough of the spot, though.

File: xarray/testing/assertions.py (hunk 9 above)
Comment (owenlittlejohns) [reply xarray/testing/assertions.py]: [970b10](https://github.com/pydata/xarray/pull/8967/commits/970b10d17cf682644031b0aa67043e7aad253ead)

File: xarray/testing/__init__.py (hunk 4 above)
Comment (flamingbear) [reply xarray/testing/__init__.py:1]: I thought there was an issue collecting things we need to do to put a final bow on things, but I'm not finding it. Should we add it to #8572? or is that overkill?

Comment (TomNicholas) [reply xarray/testing/__init__.py:1]: I can't find a dedicated issue for that either. Yes lets' just make an explicit list under **Expose datatree API publicly.** on #8572 (I'll do that now)

**Review by flamingbear — APPROVED** [review]
All of the discussion and the resulting code, looks good to me now. Thanks.

File: xarray/core/formatting.py [hunk 11]
```diff
@@ -945,6 +980,88 @@ def diff_dataset_repr(a, b, compat):
...
+
+    return "\n".join(summary)
+
+
+def diff_datatree_repr(a: DataTree, b: DataTree, compat):
+    summary = [
+        f"Left and right {type(a).__name__} objects are not {_compat_to_str(compat)}"
+    ]
+
+    # TODO check root parents?
```
Comment (TomNicholas) [inline xarray/core/formatting.py:1010]:
```suggestion
```

This should be covered by the parent assertions having a `from_root` kwarg.

File: xarray/tests/test_formatting.py [hunk 12]
```diff
@@ -555,6 +556,114 @@ def test_array_scalar_format(self) -> None:
...
+        root: DataTree = DataTree(name="root")
+        DataTree(name="results", data=dat, parent=root)
+        printout = root.__str__()
+        assert printout.splitlines()[2].startswith("    ")
+
+    def test_print_datatree(self, simple_datatree):
+        dt = simple_datatree
+        print(dt)
+
+        # TODO work out how to test something complex like this
```
Comment (TomNicholas) [inline xarray/tests/test_formatting.py:604]: This was a placeholder for something better. All this does right now is check an error is raised. But we might not need more tests than the ones already in this PR.

**Review by TomNicholas — APPROVED** [review]
A couple of very small comments about removing outdated comments / tests, after that I will merge!

File: xarray/tests/test_formatting.py (hunk 12 above)
Comment (TomNicholas) [reply xarray/tests/test_formatting.py]: Let's just pull this test, as we are already testing that the `__str__` method doesn't error in other tests.

Comment (TomNicholas) [inline xarray/tests/test_formatting.py]:
```suggestion
```

File: xarray/tests/test_formatting.py [hunk 13]
```diff
@@ -555,6 +556,112 @@ def test_array_scalar_format(self) -> None:
...
+        dat = xr.Dataset({"a": [0, 2]})
+        root: DataTree = DataTree(name="root")
+        DataTree(name="results", data=dat, parent=root)
+        printout = root.__str__()
+        assert printout.splitlines()[2].startswith("    ")
+
+    def test_print_datatree(self, simple_datatree):
+        dt = simple_datatree
+        print(dt)
+
```
Comment (TomNicholas) [inline xarray/tests/test_formatting.py:603]:
```suggestion
```
